### ISSUE TITLE
Kubeflow 1.2 installation doesn't seem to work, failing on 'secret "dns.istio-sidecar-injector-service-account" not found'

### ISSUE THREAD

**Issue by Shaked** [kubeflow/gcp-blueprints#198]
Hey, 

I have originally replied to https://github.com/kubeflow/kubeflow/issues/5048#issuecomment-785456950 with the following details. However, I believe that my issue is not directly related to https://github.com/kubeflow/kubeflow/issues/5048 so I decided to put everything in here. 

I am currently trying to install kubeflow 1.2 on GCP following https://www.kubeflow.org/docs/gke/deploy/ and I stumbled at the same problem as https://github.com/kubeflow/kubeflow/issues/5048 where DNS cloud entrypoint is not being created. However, I haven't created the service before and it doesn't exist in `gcloud --project=${PROJECT}  endpoints services list`

When I run:
```
$ kubectl -n istio-system get cloudendpoints kubeflow-demo -o yaml
```
I do see that there's a `kubeflow-demo`  object.  Having said that, it doesn't have any status.

Also when I verify the ingress:
```
kubectl -n istio-system get ingress
NAME            CLASS    HOSTS                                                      ADDRESS          PORTS   AGE
envoy-ingress   <none>   kubeflow-demo.endpoints.PROJECT.cloud.goog   XX.YYY.ZZZ.LLL   80      4h56m
```

However, when I check `gcloud` it seems like it doesn't exist:
```
gcloud --project=${KF_PROJECT} endpoints services list | grep kubeflow-demo
```

EDIT: 

I tried to deploy a completely new kubeflow instance with a new name `kubeflow-d1` just to make sure that no one has created it before, and still I end up with the same problem where the endpoint doesn't exist. 

Also, I have noticed that none of the kubeflow deployments managed to start and end up with the following error:

 > Error creating: Internal error occurred: failed calling webhook "sidecar-injector.istio.io": Post https://istio-sidecar-injector.istio-system.svc:443/inject?timeout=30s: no endpoints available for service "istio-sidecar-injector"

Running `k -n istio-system get pods` 

```
NAME                                     READY   STATUS              RESTARTS   AGE
backend-updater-0                        1/1     Running             0          10m
iap-enabler-5864d6c776-rql7t             1/1     Running             0          11m
istio-galley-7f67779964-8gbsj            0/2     ContainerCreating   0          21m
istio-galley-7f67779964-p2s2f            0/2     ContainerCreating   0          21m
istio-ingressgateway-679855b69f-4lzgk    1/2     Running             0          21m
istio-ingressgateway-679855b69f-chtcb    1/2     Running             0          21m
istio-nodeagent-kxd8n                    1/1     Running             0          21m
istio-nodeagent-m26gn                    1/1     Running             0          21m
istio-pilot-78f8495674-68gzn             1/2     Running             0          21m
istio-pilot-78f8495674-sgh4m             1/2     Running             0          21m
istio-sidecar-injector-f475c7995-4jb7q   0/1     ContainerCreating   0          21m
istio-sidecar-injector-f475c7995-lwlb9   0/1     ContainerCreating   0          21m
promsd-55f7df8c7b-g2z9n                  2/2     Running             1          21m
whoami-app-7fc9f76f57-g76mw              1/1     Running             0          18m
```

Errors: 

`k -n istio-system describe pod istio-galley-7f67779964-8gbsj` 

> MountVolume.SetUp failed for volume "dnscerts" : secret "dns.istio-galley-service-account" not found 

`k -n istio-system describe pod istio-sidecar-injector-f475c7995-4jb7q` 

> MountVolume.SetUp failed for volume "certs" : secret "dns.istio-sidecar-injector-service-account" not found 

My first thought was that none of these work because the endpoint `NAME.endpoints.PROJECT.cloud.goog` doesn't exist, but reading the again the description above made me think that `cloud-endpoints-controller` should run, or am I wrong? 

After googling for quite some time, I have found https://github.com/kubeflow/kubeflow/issues/5126, which led me to @jlewi's reply https://github.com/kubeflow/kubeflow/issues/5126#issuecomment-657728179: 

> You'll want to run a command like
`kfctl apply --context=$(KFCTXT) -f $(BUILD_DIR)/iap-ingress/ctl.isla.solutions_v1_cloudendpoint_$(NAME).yaml`

This requires [kfctl RC 1.1](https://github.com/kubeflow/kfctl/releases/tag/v1.1-rc.1) and you have to [make sure that you are logged](https://github.com/kubeflow/kubeflow/issues/2535#issuecomment-471140124) in with both commands: 

```
$ gcloud auth login
$ gcloud auth application-default login
```

Then just run `kfctl apply --context=$(KFCTXT) -f $(BUILD_DIR)/iap-ingress/ctl.isla.solutions_v1_cloudendpoint_$(NAME).yaml` as stated above. 

Even after I have solved this issue and the domain became available, the missing dns secrets issue kept on going and istio still doesn't work with the same errors as stated above. 

I also tried to run `make apply` again just to make sure I wasn't missing anything after having fixed the cloud entrypoint, but unfortunately, no luck.

I also looked into https://www.kubeflow.org/docs/gke/troubleshooting-gke/#dns-name-not-registered which ended up with no relevant information: 

```
╰─$ ku istio-system describe cloudendpoints  | grep -i status -A 5 -B 5

``` 

So basically I am looking to understand: 

1. What am I doing wrong? Obviously I either missed something in the documentation or something is just missing
2. Is there a reason why the creation of the cloud entrypoint is not automated within the `make apply`? 

Thank you
Shaked

**Comment by ihorby** [comment]
I have the same problem. Deploy kubeflow@v1.2.0  on GCP following https://www.kubeflow.org/docs/gke/deploy/deploy-cli/
And I get the same error.

`k get pods -n istio-system`
```
NAME                                     READY   STATUS              RESTARTS   AGE
istio-galley-86bd96cbf4-6hlrb            0/2     ContainerCreating   0          29m
istio-galley-86bd96cbf4-hnk4j            0/2     ContainerCreating   0          29m
istio-ingressgateway-7699694c-j4hhs      1/2     Running             0          29m
istio-ingressgateway-7699694c-pj75p      1/2     Running             0          29m
istio-nodeagent-gf5kn                    1/1     Running             0          29m
istio-nodeagent-k92ld                    1/1     Running             0          29m
istio-pilot-68d7b5b48b-pn8j5             1/2     Running             0          29m
istio-pilot-68d7b5b48b-srqmj             1/2     Running             0          29m
istio-sidecar-injector-f475c7995-js275   0/1     ContainerCreating   0          29m
istio-sidecar-injector-f475c7995-s4zch   0/1     ContainerCreating   0          29m
promsd-55f7df8c7b-twj9s                  2/2     Running             1          29m

```
`k -n istio-system describe pod istio-galley-86bd96cbf4-6hlr`
```
Warning  FailedMount  3m48s (x21 over 30m)  kubelet            MountVolume.SetUp failed for volume "dnscerts" : secret "dns.istio-galley-service-account" not found
```

`k -n istio-system describe pod istio-sidecar-injector-f475c7995-js275`
```
Warning  FailedMount  39s (x23 over 31m)   kubelet            MountVolume.SetUp failed for volume "certs" : secret "dns.istio-sidecar-injector-service-account" not found
```

**Comment by jcastagneri** [comment]
I have also seen exactly the same problem, and have similarly traced the issue to `secret "dns.istio-sidecar-injector-service-account" not found` in the Istio sidecar injector.

I also replicate that the cloudendpoint resource seems to exist for me on my management cluster, but when listing with gcloud, I don't see anything.

**Comment by jcastagneri** [comment]
## Update: I think I have a fix

### Summary
The secrets that are missing in the sidecar injector are not being created by istio because a cluster role lacks a legacy permission. Maybe an API changed underneath istio?

### Diagnosis

Diving a bit deeper into this, I found that the secrets that aren't being generated are certificates managed by istio. Istio uses [Chiron](https://istio.io/latest/docs/tasks/security/cert-management/dns-cert/) to manage certificates and autopopulate these certs as secrets. The kubeflow 1.2 blueprint configures this in the `v1_configmap_iap-ingress-config-[HASH].yaml` config under the `mesh` key. The logs for the generation of these certs are available in the istio-pilot deployment, which I grabbed with:
```
kubectl logs istio-pilot-POD-HASH -c discovery -n istio-system > /tmp/pilot-log.txt
```

The relevant failure can be seen below:
```
error   failed to approve CSR (csr-amp-ml-istio-sy-dns.isti-882tbddf6kdw9hzpws): certificatesigningrequests.certificates.k8s.io "csr-amp-ml-istio-sy-dns.isti-882tbddf6kdw9hzpws" is forbidden: user not permitted to approve requests with signerName "kubernetes.io/legacy-unknown"
error   failed to generate key and certificate for secret dns.istio-galley-service-account in namespace istio-system (error certificatesigningrequests.certificates.k8s.io "csr-amp-ml-istio-sy-dns.isti-882tbddf6kdw9hzpws" is forbidden: user not permitted to approve requests with      signerName "kubernetes.io/legacy-unknown")
```
Is the istio clusterrole underpermissioned? See the deployed clusterrole for the istio-pilot below:
```
kubectl describe clusterrole istio-pilot-istio-system                                                                                                                                                                                                                                                     [11:46:44]
Name:         istio-pilot-istio-system
Labels:       app=pilot
              release=istio
Annotations:  <none>
PolicyRule:
  Resources                                                Non-Resource URLs  Resource Names  Verbs
  ---------                                                -----------------  --------------  -----
  customresourcedefinitions.apiextensions.k8s.io           []                 []              [*]
  *.authentication.istio.io                                []                 []              [*]
  *.config.istio.io                                        []                 []              [*]
  ingresses.extensions/status                              []                 []              [*]
  ingresses.extensions                                     []                 []              [*]
  *.networking.istio.io                                    []                 []              [*]
  configmaps                                               []                 []              [create get list watch update]
  secrets                                                  []                 []              [get list watch create update delete]
  endpoints                                                []                 []              [get list watch]
  namespaces                                               []                 []              [get list watch]
  nodes                                                    []                 []              [get list watch]
  pods                                                     []                 []              [get list watch]
  services                                                 []                 []              [get list watch]
  *.rbac.istio.io                                          []                 []              [get watch list]
  *.security.istio.io                                      []                 []              [get watch list]
  certificatesigningrequests.certificates.k8s.io/approval  []                 []              [update create get delete]
  certificatesigningrequests.certificates.k8s.io/status    []                 []              [update create get delete]
  certificatesigningrequests.certificates.k8s.io           []                 []              [update create get delete]
```
I don't see anything for the legacy-unknown resource approval, so I added it:
```
- apiGroups:
  - certificates.k8s.io
  resourceNames:
  - kubernetes.io/legacy-unknown
  resources:
  - signers
  verbs:
  - approve
```

This does indeed fix the problem if I delete and redeploy the istio pilot:
```
272 2021-03-07T18:53:20.780069Z info    Istio secret "dns.istio-sidecar-injector-service-account" in namespace "istio-system" has been created
```

After this, both DNS certs for the istio sidecar and the istio galley appeared in the `istio-system` namespace.

My istio pods all recovered after this. Finally, I deleted the `kubeflow` namespace and redeployed with `make apply` to build the kubeflow resources from a clean state. This has provisioned the resources correctly (as far as I can tell!)

### Next steps
What changed with istio that would break this cluster role? Is this API drift or something else? I still don't understand why this happened.

**Comment by ihorby** [comment]
@jcastagneri 
> ## Update: I think I have a fix

Thank you, with your fix I successfully deployed the project.
I really hope that this will be fixed in the next version of Kubeflow.

**Comment by Bobgy** [comment]
Thank you @jcastagneri!
Can you report your kubernetes cluster version? That seems the most likely change.
When Kubeflow 1.2 was developed, it was against kubernetes 1.16, I believe it also works with some more, but I'm not sure the upper limit right now.

Regarding this issue, I think the root cause is still that Kubeflow 1.2 was using Anthos Service Mesh 1.4 (deprecated). @zijianjoy and I are working actively on getting latest ASM to work with Kubeflow 1.3. I think that would be a better long term solution.

**Comment by Shaked** [comment]
> I don't see anything for the legacy-unknown resource approval, so I added it:
> 
> ```
> - apiGroups:
>   - certificates.k8s.io
>   resourceNames:
>   - kubernetes.io/legacy-unknown
>   resources:
>   - signers
>   verbs:
>   - approve
> ```

This seems to work - thank you @jcastagneri . 

@Bobgy - I have tried it on `v1.18.12-gke.1210` 

I have added the following in order to automate the process, although I'm definitely not sure if this is the right way to do that: 

I am not sure if there's a way to access istio's files i.e the source of `istio/Base/Pilot/Pilot.yaml so instead, I used a different approach. 

First 
```
$ mkdir -p upstream/manifests/extras/base
$ cd upstream/manifests/extras/base
```

Then generate the following file: 

```
cat <<EOF > upstream/manifests/extras/base/kustomization.yaml
resources:
- ../../../../.build/istio/Base/Pilot/Pilot.yaml
patchesJson6902:
- target:
    group: rbac.authorization.k8s.io
    version: v1
    kind: ClusterRole
    name: istio-pilot-istio-system
  path: patch.yaml
EOF
```

```
cat <<EOF > patch.yaml
- op: add
  path: /rules/0
  value:
    apiGroups: ["certificates.k8s.io"]
    resourceNames: ["kubernetes.io/legacy-unknown"]
    resources: ["signers"]
    verbs: ["approve"]
EOF
```

It will add a merged version of Pilot.yaml and the above patch to `kubeflow-istio`

Then you can just run `make apply` and it should do the job. 

Please share if there's a better way to do that while still keeping the option to use source control and CI/CD.

Shaked

**Comment by PatrickGhosn** [comment]
Are you using a public or a private cluster? If it's a private cluster there needs to be some firewall rules added between the main and the worker nodes. We had similar issues with kfserving on a private cluster.

**Comment by jcastagneri** [comment]
@Bobgy sorry for the late reply, I am using  1.18.12-gke.1210
@Shaked thanks for the kustomize! I had similar problems finding what the true upstream file was, so I will incorporate your patch to my configuration

@PatrickGhosn  I am using a GKE cluster, but I think I'm misinterpreting what you mean by public vs private; could you help me understand?

**Comment by PatrickGhosn** [comment]
> @Bobgy sorry for the late reply, I am using 1.18.12-gke.1210
> @Shaked thanks for the kustomize! I had similar problems finding what the true upstream file was, so I will incorporate your patch to my configuration
> 
> @PatrickGhosn I am using a GKE cluster, but I think I'm misinterpreting what you mean by public vs private; could you help me understand?

![image](https://user-images.githubusercontent.com/33845447/111283304-44302680-863f-11eb-887b-8a81189361c3.png)


@jcastagneri 
Check in the cluster details. It should tell you if it's private or public like in the image above. A private cluster has extra security installed, so you will need to configure it to support KFServing. Open the right ports, configure KNative to run on the local cluster, etc...

**Comment by crmne** [comment]
@Shaked thank you for the tutorial! I tried it but didn't fix the problem unfortunately.

I also tried applying this:

```
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: istio-pilot-istio-system
  labels:
    app: pilot
    release: istio
rules:
- apiGroups:
  - certificates.k8s.io
  resourceNames:
  - kubernetes.io/legacy-unknown
  resources:
  - signers
  verbs:
  - approve
```

and then deleting the `istio-pilot`, `istio-galley`, and `istio-sidecar-injector` pods without any luck. galley and sidecar-injector are still ContainerCreating.

```
$ kubectl describe po -n istio-system istio-galley-7454849448-fcr4x
[...]
Events:
  Type     Reason       Age                  From               Message
  ----     ------       ----                 ----               -------
  Normal   Scheduled    3m36s                default-scheduler  Successfully assigned istio-system/istio-galley-7454849448-fcr4x to gke-freshflow-default-pool-f2818e64-c28m
  Warning  FailedMount  92s                  kubelet            Unable to attach or mount volumes: unmounted volumes=[dnscerts], unattached volumes=[mesh-config istio-galley-service-account-token-kwxwv envoy-config sds-uds-path istio-token dnscerts config]: timed out waiting for the condition
  Warning  FailedMount  87s (x9 over 3m35s)  kubelet            MountVolume.SetUp failed for volume "dnscerts" : secret "dns.istio-galley-service-account" not found
```

**Comment by Shaked** [comment]
> @Shaked thank you for the tutorial! I tried it but didn't fix the problem unfortunately.
> 
> I also tried applying this:
> 
> ```
> apiVersion: rbac.authorization.k8s.io/v1
> kind: ClusterRole
> metadata:
>   name: istio-pilot-istio-system
>   labels:
>     app: pilot
>     release: istio
> rules:
> - apiGroups:
>   - certificates.k8s.io
>   resourceNames:
>   - kubernetes.io/legacy-unknown
>   resources:
>   - signers
>   verbs:
>   - approve
> ```
> 
> and then deleting the `istio-pilot`, `istio-galley`, and `istio-sidecar-injector` pods without any luck. galley and sidecar-injector are still ContainerCreating.
> 
> ```
> $ kubectl describe po -n istio-system istio-galley-7454849448-fcr4x
> [...]
> Events:
>   Type     Reason       Age                  From               Message
>   ----     ------       ----                 ----               -------
>   Normal   Scheduled    3m36s                default-scheduler  Successfully assigned istio-system/istio-galley-7454849448-fcr4x to gke-freshflow-default-pool-f2818e64-c28m
>   Warning  FailedMount  92s                  kubelet            Unable to attach or mount volumes: unmounted volumes=[dnscerts], unattached volumes=[mesh-config istio-galley-service-account-token-kwxwv envoy-config sds-uds-path istio-token dnscerts config]: timed out waiting for the condition
>   Warning  FailedMount  87s (x9 over 3m35s)  kubelet            MountVolume.SetUp failed for volume "dnscerts" : secret "dns.istio-galley-service-account" not found
> ```

@crmne Can you verify that the endpoint has been created? You can see how I did it in the original issue at the top.

**Comment by crmne** [comment]
> @crmne Can you verify that the endpoint has been created? You can see how I did it in the original issue at the top.
@Shaked is this what you mean?
```
$ kubectl -n istio-system get ingress
NAME            CLASS    HOSTS                                            ADDRESS          PORTS   AGE
envoy-ingress   <none>   freshflow.endpoints.kubeflow-[redacted].cloud.goog   34.120.199.153   80      7m16s
```

**Comment by Bobgy** [comment]
Note, we are actively working on upgrading Anthos Service Mesh in https://github.com/kubeflow/gcp-blueprints/issues/144 preparing for Kubeflow 1.3 release (happening soon, rough ETA is probably a month).

Probably another good workaround can be downgrading your kubernetes cluster from regular channel to stable channel. ASM 1.4 should still work with Kubernetes 1.16. Can someone who wants to deploy now confirm that?

**Comment by zijianjoy** [comment]
> Probably another good workaround can be downgrading your kubernetes cluster from regular channel to stable channel. ASM 1.4 should still work with Kubernetes 1.16. Can someone who wants to deploy now confirm that?

I have validated that Kubernetes 1.16 deployment works fine by applying the following changes: https://github.com/kubeflow/manifests/pull/1782. We can cherrypick this for Kubeflow 1.2.1 patch.

**Comment by zijianjoy** [comment]
I have validated that GCP v1.2.1 release has been successful with https://github.com/kubeflow/gcp-blueprints/issues/208. Approving the PR and wait for website update for new deployment guide. Thank you Yuan!

**Comment by JohanWork** [comment]
@zijianjoy copied your pr into my deployment and it solved it all.

**Comment by Bobgy** [comment]
Thank you so much @zijianjoy!
